File: packages/swift-math/Sources/SwiftMath/Render/MathRenderer.swift
#if canImport(UIKit) || canImport(AppKit)
    import CoreGraphics
    import Foundation

    public enum MathRenderFailureReason: Equatable, Sendable {
        case invalidInput
        case bridgeUnavailable
        case mathJaxError
        case renderFailed
    }

    public struct MathRenderFallback: Equatable, Sendable {
        public var text: String
        public var reason: MathRenderFailureReason

        public init(text: String, reason: MathRenderFailureReason) {
            self.text = text
            self.reason = reason
        }
    }

    public enum MathRenderState {
        case pending
        case succeeded(rasterized: MathRasterizedFormula)
        case failed(fallback: MathRenderFallback)
    }

    public enum MathRenderer {
        public static func render(
            latex: String,
            options: MathRenderOptions = MathRenderOptions(),
            color: CGColor = CGColor(gray: 0, alpha: 1)
        ) -> AsyncStream<MathRenderState> {
            AsyncStream { continuation in
                continuation.yield(.pending)

                let task = Task {
                    let state = await terminalState(latex: latex, options: options, color: color)
                    guard !Task.isCancelled else {
                        return
                    }
                    continuation.yield(state)
                    continuation.finish()
                }

                continuation.onTermination = { _ in
                    task.cancel()
                }
            }
        }

        private static func terminalState(
            latex: String,
            options: MathRenderOptions,
            color: CGColor
        ) async -> MathRenderState {
            let fallbackText = fallbackText(for: latex, standalone: options.standalone)
            guard options.fontSize.isFinite, options.fontSize > 0,
                options.scale.isFinite, options.scale > 0
            else {
                return .failed(
                    fallback: MathRenderFallback(text: fallbackText, reason: .invalidInput)
                )
            }

            let bridgeOptions = MathRenderOptions(
                standalone: options.standalone,
                fontSize: options.fontSize,
                scale: 1
            )
            guard let svg = await MathJaxBridge.shared.svg(for: latex, options: bridgeOptions)
            else {
                return .failed(
                    fallback: MathRenderFallback(text: fallbackText, reason: .bridgeUnavailable)
                )
            }

            guard svg.ok else {
                return .failed(
                    fallback: MathRenderFallback(
                        text: svg.fallbackText ?? fallbackText,
                        reason: .mathJaxError
                    )
                )
            }

            guard
                let rasterized = MathRasterizer.shared.image(
                    from: svg,
                    color: color,
                    scale: CGFloat(options.scale),
                    fontSize: CGFloat(options.fontSize)
                )
            else {
                return .failed(
                    fallback: MathRenderFallback(
                        text: svg.fallbackText ?? fallbackText,
                        reason: .renderFailed
                    )
                )
            }

            return .succeeded(rasterized: rasterized)
        }

        static func fallbackText(for latex: String, standalone: Bool) -> String {
            MathJaxBridge.fallbackText(for: latex, standalone: standalone)
        }

        static func colorKey(_ color: CGColor) -> String {
            let colorSpace = color.colorSpace?.name as String? ?? "unknown"
            let components = color.components ?? []
            let componentKey = components.map(numericKey).joined(separator: ",")
            return "\(colorSpace)|\(numericKey(color.alpha))|\(componentKey)"
        }

        private static func numericKey(_ value: CGFloat) -> String {
            String(format: "%.12g", Double(value))
        }
    }
#endif

File: packages/compose-math/src/main/kotlin/io/github/dongyuzhao/composemath/MathRenderer.kt
package io.github.dongyuzhao.composemath

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class MathRenderer private constructor(private val renderSvg: suspend (String, MathRenderOptions) -> MathJaxSvg?) {
    constructor(bridge: MathJaxBridge) : this(bridge::render)

    internal constructor(service: MathRenderService) : this(service::renderSvg)

    fun render(tex: String, options: MathRenderOptions = MathRenderOptions()): Flow<MathRenderState> = flow {
        emit(MathRenderState.Pending)
        emit(renderTerminalState(tex, options))
    }

    private suspend fun renderTerminalState(tex: String, options: MathRenderOptions): MathRenderState {
        val fallbackText = MathJaxBridge.fallbackText(tex, options.displayMode)
        if (!options.fontSizeSp.isFinite() ||
            options.fontSizeSp <= 0f ||
            !options.scale.isFinite() ||
            options.scale <= 0f
        ) {
            return MathRenderState.Failed(
                MathRenderFallback(
                    text = fallbackText,
                    reason = MathRenderFailureReason.InvalidInput,
                ),
            )
        }

        val svg = renderSvg(tex, options)
            ?: return MathRenderState.Failed(
                MathRenderFallback(
                    text = fallbackText,
                    reason = MathRenderFailureReason.BridgeUnavailable,
                ),
            )

        if (!svg.ok) {
            return MathRenderState.Failed(
                MathRenderFallback(
                    text = svg.fallbackText ?: fallbackText,
                    reason = MathRenderFailureReason.MathJaxError,
                    error = svg.error,
                ),
            )
        }

        if (svg.markup.isBlank() || svg.viewBox.width <= 0f || svg.viewBox.height <= 0f) {
            return MathRenderState.Failed(
                MathRenderFallback(
                    text = svg.fallbackText ?: fallbackText,
                    reason = MathRenderFailureReason.RenderFailed,
                    error = svg.error,
                ),
            )
        }

        val document = MathSvgParser.parse(svg.markup)
        if (document == null || document.viewBox.width <= 0f || document.viewBox.height <= 0f) {
            return MathRenderState.Failed(
                MathRenderFallback(
                    text = svg.fallbackText ?: fallbackText,
                    reason = MathRenderFailureReason.RenderFailed,
                    error = svg.error,
                ),
            )
        }

        return MathRenderState.Succeeded(MathRenderedFormula(svg = svg, document = document))
    }
}
